### ISSUE TITLE
Remote Write API

### ISSUE THREAD

**Issue by roidelapluie** [prometheus/prometheus#7971]
## Proposal
**Use case. Why is this important?**

I am starting with remote_write, and from the Prometheus UI, all I can see is the URL.

I'd like new remote write users like me to easily find out in the UI:

- Does remote write work?
- What is the error message, if any?

It appears that we could also have the status of the queues: starting, resharding, ...

Let's brainstorm what would be the added value for this, and gather ideas.

Note: This is not a meta-issue; the end goal is to have a remote-write page in the UI.

cc @cstyan @callumj @bwplotka @juliusv

**Comment by cstyan** [comment]
Should be simple enough to display some of the common metrics for each queue on a page.

**Comment by csmarchbanks** [comment]
Detail on the sharding calculations would be interesting, right now that information is only available in debug logs.

**Comment by cstyan** [comment]
@roidelapluie this feels like it could be a hacktoberfest task?

**Comment by roidelapluie** [comment]
Yes, indeed :)

**Comment by Pokom** [comment]
Would this be achievable to a new contributor? I'd like to take a shot at it once there's enough requirements.

**Comment by cstyan** [comment]
@Pokom I would think so, I haven't done any work with any of the UI pages but it looks like the source is [here](https://github.com/prometheus/prometheus/tree/8c2bc2f57a4e01121dbe0377fbaa7805311afc8e/web/ui/react-app/src/pages), and you'd either just be pulling existing data from remote write or filing issues/opening additional PR's to make unavailable info available to the UI page.

**Comment by roidelapluie** [comment]
Like this @cstyan @csmarchbanks ?

![queues](https://user-images.githubusercontent.com/291750/96173887-cbf6f500-0f28-11eb-9fcf-f0791776507f.png)

**Comment by csmarchbanks** [comment]
I like the state of each shard, that could have some good information, especially the pending samples per shard. I am curious, what do last scrape and scrape duration mean?

Generally, I think a section for information about the queue is also necessary, the whole queue is either resharding or running, not individual shards. Each shard would be running, stopping, or stopped I think.

**Comment by cstyan** [comment]
Might be nice to have a collapsible section for each overall queue that shows the queue config and shard information in a table?

**Comment by roidelapluie** [comment]
> I like the state of each shard, that could have some good information, especially the pending samples per shard. I am curious, what do last scrape and scrape duration mean?
> 
> Generally, I think a section for information about the queue is also necessary, the whole queue is either resharding or running, not individual shards. Each shard would be running, stopping, or stopped I think.

Yeah should be "last sent timestamp" and "last sent duration"

**Comment by csmarchbanks** [comment]
> Might be nice to have a collapsible section for each overall queue that shows the queue config and shard information in a table?

I would say collapse the shard information by default, but if the queue information is reasonably compact and high level then that could remain prominently displayed for a quick scan?

**Comment by strideynet** [comment]
Has this task been picked up by someone?

**Comment by roidelapluie** [comment]
> Has this task been picked up by someone?

No, it is free to pick :)

**Comment by strideynet** [comment]
I've picked this issue up, hoping I will have time to finish it in the next few days for review.

**Comment by roidelapluie** [comment]
If you want to share a mock before doing the hard work, that would be welcome. The exemple I provided will need improvements.

**Comment by roidelapluie** [comment]
And thanks for picking this :)

**Comment by strideynet** [comment]
Just to confirm, I'm needing to add a new endpoint for this data to be accessible via the API. There's no issue introducing that with the same PR that implements the web UI?

**Comment by brian-brazil** [comment]
It'd be unusual not to.

**Comment by strideynet** [comment]
First rough mock out of the UI just throwing together some react components raises a few questions. It's a rough mock but any feedback on this would be appreciated.

![image](https://user-images.githubusercontent.com/16336790/99919717-02d5dd00-2d17-11eb-96fa-36f852395c43.png)

1. For the values used as part of the sharding calculations we have two choices: we can show the values as they are currently, or as they were the last time the sharding calculation was run to calculate a desired value. Since this runs every ten seconds, it doesn't seem to make a huge difference, but we obviously need to make a choice here.

2. Within the QueueManager there's currently no concept of a "state" for the shards and queue. I can't tell whether or not it is worth introducing this in order to service this UI? The only value I think that would add significant value is a "Resharding" status shown somewhere for the queue.

**Comment by roidelapluie** [comment]
That looks nice. I fell the error message could be in red like in the targets page. Also, I would expect the name of the remote read as title and the URL as part of the config displayed.

1. I would pick the current value, but I let that to remote write maintainers
2. The state would be either OK or error if there is an error. That would indeed only be cosmetic.

**Comment by csmarchbanks** [comment]
Nice work, thanks for putting this together!

 First, to answer your questions:

1. I think last calculation run is nice along with the time in which it was run, but don't have strong opinions either way.
2. I agree with roidelapluie as to state, we just need some indicator as to whether the last request was an error or not.

A couple of additional comments/thoughts:

1. I would like the name of the remote write and the URL in the title, otherwise if it is an unnamed queue the has by itself is not very useful. You might already have this if "Primary Remote" is the name, but I want to make sure :)
2. The shard info could be moved down into the calculation area, I think a more interesting headline would be the delay and status.

**Comment by roidelapluie** [comment]
Note: my expectation is that 99% of remote write users would have just one remote, right? That might help in the design to think about that use case first (when it comes to names etc).

**Comment by strideynet** [comment]
> Nice work, thanks for putting this together!
> 
> First, to answer your questions:
> 
> 1. I think last calculation run is nice along with the time in which it was run, but don't have strong opinions either way.
> 2. I agree with roidelapluie as to state, we just need some indicator as to whether the last request was an error or not.
> 
> A couple of additional comments/thoughts:
> 
> 1. I would like the name of the remote write and the URL in the title, otherwise if it is an unnamed queue the hash by itself is not very useful. You might already have this if "Primary Remote" is the name, but I want to make sure :)
> 2. The shard info could be moved down into the calculation area, I think a more interesting headline would be the delay and status.

Yeah, the title is currently Name and then the Endpoint. I observed the hash behaviour myself last week :)

I think I agree with your suggestion regarding the shard info since most of those figures aren't immediately eye-catching, delay + status makes a bit more sense here.

I've started with pulling the data through from the QueueManager to the API, but feel like I may need to chat with you on irc sometime in the next few days to confirm some bits since I'm not too sure on the house preferences .

I'm in the middle of preparing to transition to a new role with a new employer, so I may need to take a short break from this task.

**Comment by roidelapluie** [comment]
Would it be worth it to already open the PR?

**Comment by strideynet** [comment]
https://github.com/prometheus/prometheus/pull/8218

Perhaps move discussion here.

**Comment by Krishna-Sivakumar** [comment]
Is this issue still open? I'd like to take it up.

**Comment by cstyan** [comment]
Yes. Might be a good idea to take a look at the previously opened PR. https://github.com/prometheus/prometheus/pull/8218

**Comment by shashank-priyadarshi** [comment]
Hi @roidelapluie if this issue is still open, I can take this up!

**Comment by cstyan** [comment]
AFAIK no one has worked on this, so feel free to. The previous attempt at an implementation plus the above comments should help get you started.

**Comment by kushalShukla-web** [comment]
Hey guys i saw there is no PR in this issue and i would liked to work on it .

**Comment by kushalShukla-web** [comment]
hey hi @juliusv  i would love to work on this new UI of prometheus , where can i find the issues for these new UI ?

**Comment by juliusv** [comment]
@kushalShukla-web Sorry for the late reply! We don't have specific issues for the new UI yet, but I would just consider any non-urgent UI feature request an issue for the new UI as of now :)

**Comment by vijaygovindaraja** [comment]
hey, i'd like to take a shot at this. have a working implementation with:

- new `GET /api/v1/status/remote_write` endpoint that returns per-queue status (sharding, pending samples, sent/failed/retried counts, delay, bytes sent)
- new Mantine UI page under Status > Remote write showing health status, delay badges, shard info, and detailed metrics per queue
- backend reads directly from QueueManager metrics with proper locking
- tests for the new status API

will open a PR shortly.

**Comment by vijaygovindaraja** [comment]
PR is up: #18388

@roidelapluie @juliusv @csmarchbanks would appreciate a look when you get a chance
